### ISSUE TITLE
[Http-Client] Partial response returned on "max_duration" when using Psr18Client wrapper 

### ISSUE THREAD

**Issue by vbartusevicius** [symfony/symfony#52490]
### Symfony version(s) affected

5.4.26

### Description

When "max_duration" option is passed to any of `HttpClientInterface` implementations AND those are wrapped to `Psr18Client`, a partial response is returned instead of an error when "max_duration" is hit.

### How to reproduce

We need a `server.php` that returns the response slowly:

```php
<?php

set_time_limit(0);
ob_implicit_flush(1);

$json = <<<'JSON'
{"here": "is","some": "json","with": "a","lot": "of","keys": "and","values": 42}
JSON;

header('Content-Type: application/json');

foreach (str_split($json) as $letter) {
    echo $letter;
    usleep(0.1 * 1000000);
}
```

The `client.php` makes two calls to the "server" - one with plain `CurlHttpClient`, and another with `Psr18Client` wrapped around:

```php
<?php

include __DIR__ . '/vendor/autoload.php';

use Symfony\Component\HttpClient\CurlHttpClient;
use Symfony\Component\HttpClient\Psr18Client;
use Symfony\Contracts\HttpClient\HttpClientInterface;

$curlClient = new CurlHttpClient([
    'timeout' => 0.25,
    'max_duration' => 0.75,
]);
$url = 'http://slow-server:3064/';

echo "Plain CurlHttpClient:\n";
usePlainCurl($curlClient, $url);
echo "\n\n";
echo "PSR-18 CurlHttpClient:\n";
usePsr18Curl($curlClient, $url);

function usePlainCurl(HttpClientInterface $client, string $url): void
{
    $response = $client->request('GET', $url);

    try {
        echo sprintf("STATUS: %s\n", $response->getStatusCode());
        echo sprintf("BODY: %s\n", $response->getContent(false));
    } catch (Exception $exception) {
        echo sprintf('ERROR: %s', $exception->getMessage());
    }
}

function usePsr18Curl(HttpClientInterface $client, string $url): void
{
    $psrClient = new Psr18Client($client);
    $request = $psrClient->createRequest('GET', $url);

    try {
        $response = $psrClient->sendRequest($request);

        echo sprintf("STATUS: %s\n", $response->getStatusCode());
        echo sprintf("BODY: %s\n", $response->getBody()->getContents());

    } catch (Exception $exception) {
        echo sprintf('ERROR: %s', $exception->getMessage());
    }
}
```

Running `php client.php` results in a partial response returned by the PSR18 client with no error:
```bash
bash-5.1# php client.php 
Plain CurlHttpClient:
STATUS: 200
ERROR: Operation timed out after 753 milliseconds with 8 bytes received for "http://slow-server:3064/".

PSR-18 CurlHttpClient:
STATUS: 200
BODY: {"here
```

Part of the used `docker-compose.yml`:
```yaml
  slow-server:
    container_name: slow-server
    image: php:8.2-cli
    command: ["php", "-S", "0.0.0.0:3064", "/usr/src/server.php"]
    volumes:
      - ./server.php:/usr/src/server.php
    ports:
      - "3064:3064"
```

### Possible Solution

No partial body is returned with the `Response`. An exception is thrown when "max_duration" is hit while using the `Psr18Client` wrapper.

**Comment by nicolas-grekas** [comment]
After investigating this issue a bit, I figured out that the issue is on the side of the PSR-7 implementation.
If you're using nyholm/psr7, I submitted a fix at https://github.com/Nyholm/psr7/pull/252
Other PSR-7 implementations might need a similar fix.
Feel free to send them PRs that reuse my code and link back to https://github.com/Nyholm/psr7/pull/252

**Comment by nicolas-grekas** [comment]
guzzle/psr7 looks OK BTW:
https://github.com/guzzle/psr7/blame/2.6/src/Utils.php#L404

**Comment by nicolas-grekas** [comment]
Fixed in nyholm/psr7 v1.8.1
